### DISCUSSION TITLE
Store HA config entries as a global and update on change

### DISCUSSION THREAD

**Discussion by mdegat01** [zachowj/node-red-contrib-home-assistant-websocket#385] [Feature Request]
Part cookbook, part feature request. I recently found myself dealing with a few device actions from my nest integration and trying to make node red flows which trigger off them. The problem is the events fired for device actions only have the device ID in them and there's no easy way to determine which entities are tied to that device ID to look at their state information. So I ended up making this crappy manual map in my function node like this:

```js
context.set("devices",{
    "2b4b97d69028e9ebdb3af5f9c9be6bc0": "camera.family_room",
    "77c761c57a60f494ad2be5eeec9c75d1": "camera.kitchen",
    "e96aa8792946fd5bcc555dab0f5f2868": "camera.porch",
    "a409647172dee6d6c2096f184f7c6f92": "climate.dining_room"
});
```

It works but its ugly and fragile. You have to look up every id from `.storage` and copy it in. And then if you ever need to recreate the integration from scratch then it breaks since the IDs are just randomly generated.

So I set to work on a better solution. I discovered there are events fired you can listen for whenever the device, area or entity registries are updated (`area_registry_updated`, `device_registry_updated` and `entity_registry_updated`). And the websockets API includes a way to pull each of the registries by sending the message `{"type":"config/area|device|entity_registry/list"}`.

The final registry (config entries) is different by also accessible. There's a REST API - `/api/config/config_entries/entry`. There's no event for when its updated I could find unfortunately but I found what worked pretty well was also updating it when I received an event that the device or event registries were updated. Listening for a change in services would be even more complete (since its possible for an integration to only add services).

I then made a subflow out of this which listens for these events and stashes the registries in the `homeassistant.config` global. This way I can filter through it to look things up as part of automations on demand and its always up to date:
```json
[{"id":"16fdbd81.9d9b82","type":"subflow","name":"Store HA config","info":"Keep config of HA up to date and easily accessible in globals. That includes:\n\n- Config entries\n- Areas\n- Devices\n- Entities\n\nIt also flattens areas into entities. So any time devices or areas are updated it also updates entities by filling out their `area_id` field. Normally entities have this blank and instead expect the device to set the area.","category":"","in":[{"x":160,"y":40,"wires":[{"id":"b09b8d39.fe41e"},{"id":"bd3c0c61.c6069"},{"id":"9349652e.f4a0c"}]}],"out":[{"x":1120,"y":160,"wires":[{"id":"447143e7.58fbf4","port":0}]}],"env":[],"color":"#DDAA99","status":{"x":240,"y":360,"wires":[{"id":"4e6fd6ff.adeda8","port":0}]}},{"id":"7b548b89.a5c924","type":"ha-api","z":"16fdbd81.9d9b82","name":"Areas","server":"cc03735a.94933","debugenabled":false,"protocol":"websocket","method":"get","path":"","data":"{\"type\":\"config/area_registry/list\"}","dataType":"json","location":"payload.areas","locationType":"msg","responseType":"json","x":370,"y":160,"wires":[["447143e7.58fbf4"]]},{"id":"bd3c0c61.c6069","type":"ha-api","z":"16fdbd81.9d9b82","name":"Devices","server":"cc03735a.94933","debugenabled":false,"protocol":"websocket","method":"get","path":"","data":"{\"type\":\"config/device_registry/list\"}","dataType":"json","location":"payload","locationType":"msg","responseType":"json","x":320,"y":220,"wires":[["e9500159.2c764"]]},{"id":"b09b8d39.fe41e","type":"ha-api","z":"16fdbd81.9d9b82","name":"Entities","server":"cc03735a.94933","debugenabled":false,"protocol":"websocket","method":"get","path":"","data":"{\"type\":\"config/entity_registry/list\"}","dataType":"json","location":"payload","locationType":"msg","responseType":"json","x":320,"y":280,"wires":[["418cca4a.cbe594"]]},{"id":"dc8fe39c.fde6d8","type":"ha-api","z":"16fdbd81.9d9b82","name":"Config entries","server":"cc03735a.94933","debugenabled":false,"protocol":"http","method":"get","path":"/api/config/config_entries/entry","data":"{}","dataType":"json","location":"payload.entries","locationType":"msg","responseType":"json","x":760,"y":220,"wires":[["447143e7.58fbf4"]]},{"id":"2c4c713b.bb30ae","type":"server-events","z":"16fdbd81.9d9b82","name":"Entity reg updated","server":"cc03735a.94933","event_type":"entity_registry_updated","exposeToHomeAssistant":false,"haConfig":[{"property":"name","value":""},{"property":"icon","value":""}],"waitForRunning":true,"x":110,"y":280,"wires":[["b09b8d39.fe41e"]]},{"id":"9633cbe3.e13cd8","type":"server-events","z":"16fdbd81.9d9b82","name":"Device reg updated","server":"cc03735a.94933","event_type":"device_registry_updated","exposeToHomeAssistant":false,"haConfig":[{"property":"name","value":""},{"property":"icon","value":""}],"waitForRunning":true,"x":110,"y":220,"wires":[["bd3c0c61.c6069"]]},{"id":"1f22fbb3.0144d4","type":"server-events","z":"16fdbd81.9d9b82","name":"Area reg updated","server":"cc03735a.94933","event_type":"area_registry_updated","exposeToHomeAssistant":false,"haConfig":[{"property":"name","value":""},{"property":"icon","value":""}],"waitForRunning":true,"x":100,"y":160,"wires":[["9349652e.f4a0c"]]},{"id":"447143e7.58fbf4","type":"function","z":"16fdbd81.9d9b82","name":"Update global","func":"const data = msg.payload;\n\nif(data.entries){\n    update_config_entries(data.entries);\n    node.status({text:`Updated config entries (${data.entries.length})`});\n}\nif(data.areas){\n    update_areas(data.areas);\n    node.status({text:`Updated areas (${data.areas.length})`});\n}\nif(data.devices){\n    update_devices(data.devices);\n    node.status({text:`Updated devices (${data.devices.length})`});\n}\nif(data.entities){\n    update_entities(data.entities);\n    node.status({text:`Updated entities (${data.entities.length})`});\n}\n\nreturn msg;\n\nfunction update_entities(entities){\n    if(!entities){\n        entities = global.get('homeassistant.config.entities') || [];\n    }\n    \n    devices = global.get('homeassistant.config.entities');\n    for (let e of entities){\n        if(!e.area_id && e.device_id){\n            e.area_id = devices.find(d => d.id == e.device_id) || null;\n        }\n    }\n    \n    global.set('homeassistant.config.entities', entities);\n}\n\nfunction update_devices(devices){\n    global.set('homeassistant.config.devices', devices);\n    update_entities();\n}\n\nfunction update_areas(areas){\n    global.set('homeassistant.config.areas', areas);\n    update_entities();\n}\n\nfunction update_config_entries(entries){\n    global.set('homeassistant.config.entries', entries);\n}","outputs":1,"noerr":0,"initialize":"","finalize":"","x":1000,"y":160,"wires":[[]]},{"id":"fc6ceac.c313e98","type":"join","z":"16fdbd81.9d9b82","name":"Make payload","mode":"custom","build":"object","property":"payload","propertyType":"msg","key":"topic","joiner":"\\n","joinerType":"str","accumulate":false,"timeout":"1","count":"","reduceRight":false,"reduceExp":"","reduceInit":"","reduceInitType":"","reduceFixup":"","x":580,"y":220,"wires":[["dc8fe39c.fde6d8"]]},{"id":"e9500159.2c764","type":"change","z":"16fdbd81.9d9b82","name":"Set topic","rules":[{"t":"set","p":"topic","pt":"msg","to":"devices","tot":"str"}],"action":"","property":"","from":"","to":"","reg":false,"x":415,"y":220,"wires":[["fc6ceac.c313e98"]],"l":false},{"id":"418cca4a.cbe594","type":"change","z":"16fdbd81.9d9b82","name":"Set topic","rules":[{"t":"set","p":"topic","pt":"msg","to":"entities","tot":"str"}],"action":"","property":"","from":"","to":"","reg":false,"x":415,"y":280,"wires":[["fc6ceac.c313e98"]],"l":false},{"id":"4e6fd6ff.adeda8","type":"status","z":"16fdbd81.9d9b82","name":"","scope":["447143e7.58fbf4"],"x":140,"y":360,"wires":[[]]},{"id":"9349652e.f4a0c","type":"change","z":"16fdbd81.9d9b82","name":"Clear payload","rules":[{"t":"set","p":"payload","pt":"msg","to":"{}","tot":"json"}],"action":"","property":"","from":"","to":"","reg":false,"x":275,"y":160,"wires":[["7b548b89.a5c924"]],"l":false},{"id":"69ecad84.cd05f4","type":"server-events","z":"16fdbd81.9d9b82","name":"HA client","server":"cc03735a.94933","event_type":"home_assistant_client","exposeToHomeAssistant":false,"haConfig":[{"property":"name","value":""},{"property":"icon","value":""}],"waitForRunning":true,"x":80,"y":100,"wires":[["a31c45ce.09ba98"]]},{"id":"a31c45ce.09ba98","type":"switch","z":"16fdbd81.9d9b82","name":"Connected event","property":"payload","propertyType":"msg","rules":[{"t":"eq","v":"connected","vt":"str"}],"checkall":"true","repair":false,"outputs":1,"x":175,"y":100,"wires":[["9349652e.f4a0c","bd3c0c61.c6069","b09b8d39.fe41e"]],"l":false},{"id":"cc03735a.94933","type":"server","name":"Home Assistant","legacy":false,"addon":true,"rejectUnauthorizedCerts":true,"ha_boolean":"y|yes|true|on|home|open","connectionDelay":true,"cacheJson":true}]
```
![image](https://user-images.githubusercontent.com/2037026/117194754-76921080-adb2-11eb-9841-bf241ab621d3.png)

And this does seem to work great for anyone interested. But I guess what I was wondering is can this just be a feature of this module? Having up to date access to this information seems handy at times for things like reloading config entries, taking an action on everything in an area or handling device actions.

**Comment by zachowj** [comment]
I haven't fully read through your post and this doesn't answer it directly but for your use case of the Nest device action have you taken a look at the [beta](https://github.com/zachowj/node-red-contrib-home-assistant-websocket/discussions/382) with the new device node?

**Reply by mdegat01** [reply]
I haven't yet, that does sound good though! How do I install the beta if I use the add-on? Put that specific version in the list of `npm_packages` in the add-on config?

Although I still think there are other uses to that info. Like being able to find all entities in an area and then act on them in a flow. Or is that something that's also in the works for the `get entities` node?

Btw, what do you actually store in the node for device actions? The device ID like HA does? Obviously that works but is still a bit fragile for the reasons I mentioned above, if you have to remove and re-add the integration or something then you get all new device IDs. That's why I like being able to filter and find the ID by name since I can control the name and always call my devices the same things.

**Reply by zachowj** [reply]
> How do I install the beta if I use the add-on?

The very bottom of the post https://github.com/zachowj/node-red-contrib-home-assistant-websocket/discussions/382

> Btw, what do you actually store in the node for device actions?

Yes, it uses the exact same format as a HA device automation because it uses the same backend but triggers an NR node instead of a HA automation.

> Although I still think there are other uses to that info.

There are definitely use-cases for having access to that information. Though not sure it would be used by the majority. It's the same concept as my cookbook entry for [state_changed event based on area](https://zachowj.github.io/node-red-contrib-home-assistant-websocket/cookbook/get-state_changed-events-based-on-area.html).

I have been meaning to add an option to turn off saving all entities to the global context because of possible security issues as any node has access to it and there could be sensitive information in the entities list. Adding more to the global context could possibly open a bigger window into HA.

> Like being able to find all entities in an area and then act on them in a flow.

Not sure this fits your need but the new target stuff in the call-service node allows you to act on all entities in an area, device, etc...

> Or is that something that's also in the works for the `get entities` node?

I think this would be the approach to take. Adding options to the get-entities node to be able to find entities by device and areas.

**Reply by mdegat01** [reply]
> The very bottom of the post #382

🤦 My bad, I see it now. Ok I'll check it out then.

> Not sure this fits your need but the new target stuff in the call-service node allows you to act on all entities in an area, device, etc...

I'll check that out too, don't think I had seen that.

> I have been meaning to add an option to turn off saving all entities to the global context because of possible security issues as any node has access to it and there could be sensitive information in the entities list. Adding more to the global context could possibly open a bigger window into HA.

😱 <- This is me realizing this may become deprecated.

I see your point but omg, the amount of work I'd need to do... I really hated how complicated it made my flows to constantly put get state nodes everywhere before every switch/change/function node so I basically just stopped doing it. I rarely use that node and instead just use `$globalContext` or `global.get` whenever I need HA state data. I can't even imagine how much refactoring work I'd have to do to get rid of all that 😩

Could the `$entity` jsonata function be made available everywhere instead of only in the HA nodes or is that not possible? If that could be made available everywhere then I wouldn't really mind. I'd still have to do a bunch of refactoring but at least all the jsonata would be much cleaner in the end, having `$globalContext("homeassistant.homeAssistant.states['<entity id>'].state")` everywhere is pretty ugly.

**Reply by zachowj** [reply]
> This is me realizing this may become deprecated.

I have no plan to remove it but I'll probably move it to opt-in for new users.
